Make multi-valued input fields more user-friendly

kiersten-stokes opened the pull request:
Fixes #2750
Fixes #2738
Fixes #2923 
Fixes #2906

Corresponds to elyra-ai/pipeline-editor#201.

Excuse all the extraneous commits - I was building and merging with the `pipeline-rjsf` branch so often and for so long that I didn't want to mess with the history by cherry-picking. The diff, however, is accurate.

Note that the pipeline editor may go blank when trying to edit multi-valued fields at this point during development. Front end changes required before things are fully functional. 

### What changes were proposed in this pull request?
This PR re-factors a significant portion of backend node and pipeline properties handling in order to centralize and simplify validation and processing of Elyra-owned properties (which includes properties such as `mounted_volumes`, `kubernetes_secrets`, etc. that are not properties considered a part of the operator/component definition itself).


TODOs:
- [x] **frontend**: pipeline editor goes blank
- [x] **frontend**: tooltip for airflow nodes doesn't render properties properly
- [x] **frontend**: update CSS/styling of new input format
- [x] **frontend**: update default property propagation for new format
- [x] **frontend**/**backend**: integrate UI placeholders (or decide on a title) for each input field
- [x] **frontend**: update property display on node mouseover
- [x] **frontend**: remove/do not persist properties with all empty values (may be tackled in followup since backend validation will catch)
- [x] **frontend**: fix env var parsing and refresh for new format
- [x] ~**frontend**: frontend validation (may require schema updates on backend)~ backend validation will handle most cases; frontend validation can be added in a follow-up if required
- [x] ~**frontend**/**backend**: investigate improving display of Component Source property (#2256) (may be better tackled in a follow-up PR)~ non-trivial, will have to tackle in a follow-up
- [x] **frontend**: migration
- [x] **documentation** Update [node properties reference topic in pipelines topic](https://elyra.readthedocs.io/en/latest/user_guide/pipelines.html#node-properties-reference) (for many properties we currently discuss the string input format)
- [x] pipeline editor release + merge


### How was this pull request tested?
- [x] Fix backend tests
- [x] Fix integration tests
- [x] Reviewed output of `make docs` (section "Node properties reference" in the pipelines documentation topic)

Manual test checklist. If checked, end-to-end test was successful:
 - node properties are rendered correctly (this includes validation for missing and invalid property input)
 - node tooltip on canvas renders properties correctly
 - properties are persisted as expected  in `.pipeline` file
 - properties are persisted as expected in exported DAG
 - properties are properly applied to pod 

pipeline properties:
  - [x] data volumes
  - [x] annotations
  - [x] environment variables
  - [x] tolerations
  - [x] secrets
  - [x] disable node caching
 
node properties (KFP)
  - [x] data volumes - generic components
  - [x] data volumes - custom components
  - [x] annotations - generic components
  - [x] annotations - custom components
  - [x] environment variables - generic components
  - [x] tolerations - generic components
  - [x] tolerations - custom components
  - [x] secrets - generic components
  - [x] disable node caching - custom components
  
node properties (Airflow)
  - [x] data volumes - generic components
  - [x] data volumes - custom components
  - [x] annotations - generic components
  - [x] annotations - custom components
  - [x] environment variables - generic components
  - [x] tolerations - generic components
  - [x] tolerations - custom components
  - [x] secrets - generic components


 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

ptitzler commented:
![image](https://user-images.githubusercontent.com/13068832/191185653-2b55d42b-faf6-49e5-b2a4-088d3f53032e.png)


If one chooses (or leaves) the default selection, this is persisted in the `.pipeline` file:
```
"disallow_cached_output": "Use runtime environment default",
```

Since `Use runtime environment default` is a UI label that could change over time, it'd be safer to persist `null` instead to indicate that the user did not choose an explicit property value.

ptitzler commented:
![image](https://user-images.githubusercontent.com/13068832/191186584-14bc4b1b-672a-4b71-ac89-7da991a87546.png)

Console log:
```
3935.4159b022aa6d82e44127.js?v=4159b022aa6d82e44127:2 TypeError: Cannot set properties of undefined (setting 'ui:placeholder')
    at index.js:1:37929
    at Proxy.forEach (<anonymous>)
    at p (index.js:1:37718)
    at index.js:1:38167
    at Immer.produce (immer.cjs.development.js:792:1)
    at t.propagatePipelineDefaultProperties (index.js:1:37339)
    at t.getAllPaletteNodes (index.js:1:37213)
    at index.js:1:67826
    at Nu (3935.4159b022aa6d82e44127.js?v=4159b022aa6d82e44127:2:61501)
    at uo (3935.4159b022aa6d82e44127.js?v=4159b022aa6d82e44127:2:69686)
```

ptitzler commented on elyra/pipeline/component_parameter.py:
```diff
@@ -0,0 +1,818 @@
+    _keys = ["path"]
+    _ui_details_map = {
+        "path": {"display_name": "Mount Path", "placeholder": "/mount/path"},
+        "pvc_name": {"display_name": "Volume Claim Name", "placeholder": "pvc-name"},
```
```suggestion
        "pvc_name": {"display_name": "Persistent Volume Claim Name", "placeholder": "pvc-name"},
```

ptitzler commented:
Generic nodes: the `runtime image` property is now rendered as the last property in the list, even though it is the most important one:

Pipeline properties:
![image](https://user-images.githubusercontent.com/13068832/191192709-fbb8254f-9f04-4e27-8d2c-626c58d74dab.png)


Node properties:
![image](https://user-images.githubusercontent.com/13068832/191191944-97018a14-94d0-4058-b5ad-cf0520c8d796.png)

ptitzler commented:
Validation doesn't detect missing annotation value

![image](https://user-images.githubusercontent.com/13068832/191194294-b33d923d-bc24-4c0c-b105-2ef2cdd4e830.png)
As a result the invalid key/value combination is persisted in the `.pipeline` file and exported, producing an invalid annotation.

ptitzler commented:
Environment variables widget:

- If no value is specified, a cryptic validation error is raised:

  ![image](https://user-images.githubusercontent.com/13068832/191209116-63729309-9f40-4a6a-9d18-d5228cbd7935.png)

  ```
  Property has an improperly formatted env variable key value pair.
  ```

  In previous releases an empty string was considered valid. I believe we have to retain that behavior because otherwise the environment variable detection code ("refresh" button) would produce invalid list entries if the analyzed source code does not set a default value.

 - "Refresh" button doesn't work - nothing is displayed in the env variable list
 - spaces in environment variable names should not be allowed. while it might be possible to process such variables programmatically, it's generally a major pain in shell environments/scripts

ptitzler commented:
`data volumes` widget:

- Validation error metadata is inconsistent across widgets. One would expect the same validation error to be raised if a constraint violation is detected. For example, assume widget X requires input of a key/value pair and widget Y also requires input of a key/value pair. If the user doesn't provide a value, validation should return messages that are consistent. For data volumes the validation error metadata is 
   
![image](https://user-images.githubusercontent.com/13068832/191264746-c579abdc-9382-4b56-9089-f396f1ba0618.png)

Note the use of `"<missing required value>"`

kevin-bates commented on elyra/pipeline/processor.py:
```diff
@@ -352,15 +366,56 @@ def _sort_operation_dependencies(operations_by_id: dict, ordered_operations: lis
                     )
             ordered_operations.append(operation)
 
+    def add_runtime_image(self, instance, execution_object: Any, **kwargs) -> None:
```
Should these be on `RuntimePipelineProcessor` instead since they don't apply to local execution?

Could we use a forward reference to annotate the type on `instance`?  Would a name like `runtime_processor` be more clear?
```suggestion
    def add_runtime_image(self, runtime_processor: "RuntimePipelineProcessor", execution_object: Any, **kwargs) -> None:
```

kiersten-stokes replied:
In this case the instance refers to the `ElyraProperty` (since each individual property calls `runtime_processor.add_{property_name}(self, execution_object`). But I've gone ahead and type-hinted those anyway since it's probably for the best 😄

kiersten-stokes commented:
@ptitzler thanks again for the review! I believe I've addressed everything that I could from the above comments. Details below


> Since `Use runtime environment default` is a UI label that could change over time, it'd be safer to persist null instead to indicate that the user did not choose an explicit property value.

I've changed this such that only `True` and `False` are listed. The empty string is persisted, which is handled during property conversion in the `DisallowCachedOutput` class (though full disclosure, I haven't been able to thoroughly test this due to the pipeline editor white-outs).

> Generic nodes: the runtime image property is now rendered as the last property in the list, even though it is the most important one

I've removed the `RuntimeImage` `ElyraProperty` class and added the property manually in the generic and pipeline properties templates (as it was before), restoring it to it's rightful place in the property panel 😄 

> Validation doesn't detect missing annotation value

Addressed and added tiny test case to the existing test for invalid annotations.

<img width="464" alt="Screen Shot 2022-09-20 at 3 06 31 PM" src="https://user-images.githubusercontent.com/31816267/191354251-e2878996-76a1-4b5d-b169-536c5e772f0b.png">

> [environment variables] Need to add placeholder values (ENV_VAR, value)that are consistent with other widgets

I haven't quite been able to figure this one out. The placeholders exist in the properties JSON the same way as for the others but aren't showing up. @marthacryan (or @ajbozarth) I think I might need your help on this one. [Here's](https://github.com/elyra-ai/elyra/pull/2927#issuecomment-1252038890) the originating comment.

> if no key/value pair is enters a fatal error occurs during validation

Fixed for all properties!

> [environment variables] If no value is specified, a cryptic validation error is raised:

Now when no value is specified, instances are silently removed from the list (same behavior as current `main`).

> "Refresh" button doesn't work - nothing is displayed in the env variable list

Yeah this has been on my radar, and I believe after looking into it that it is a frontend task - tagging @ajbozarth since I think he implemented the functionality. [Here's](https://github.com/elyra-ai/elyra/pull/2927#issuecomment-1252038890) the originating comment.

> spaces in environment variable names should not be allowed. 

Fixed with below message and added test case to existing test

<img width="661" alt="Screen Shot 2022-09-20 at 3 16 05 PM" src="https://user-images.githubusercontent.com/31816267/191356005-87662047-f2c9-4711-992a-4e28f863fa98.png">

> no validation error is raised if only a pvc name is specified, but not a mount path

Fixed with below message and added test case to existing test

<img width="418" alt="Screen Shot 2022-09-20 at 3 23 27 PM" src="https://user-images.githubusercontent.com/31816267/191357302-c44c5c90-979c-4aad-81f3-5d0e23c99c10.png">

> validation - replace <missing value> with what the actual entered value is

Done! As seen in the above screenshots

> [kubernetes tolerations] key is not a required property. the previously implemented validation code should already have the complete logic built that verifies the constraints that need to be enforced

Empty key, value, and effect are now all allowed in the correct circumstances

> [kubernetes tolerations] the initial selected value for operator should be Equal to properly model the Kubernetes behavior
> [kubernetes tolerations] operator property is not properly recognized (note Exists -> missing required value)

I've updated this on the backend such that "Equal" is set as the default in the schema. I believe the second part, however, will require a frontend change. Or at least feedback as to what I might need to change in the schema. Basically, the value is only persisted once it's changed for the first time. @marthacryan let me know if there's something I could change to fix this! Originating comment [here](https://github.com/elyra-ai/elyra/pull/2927#issuecomment-1252358337)

> effect property should be an select list enum ["", "NoExecute", "NoSchedule", "PreferNoSchedule"]

Done

<img width="380" alt="Screen Shot 2022-09-20 at 3 31 05 PM" src="https://user-images.githubusercontent.com/31816267/191358525-da15bd1b-3d6a-4327-8e1d-f47b489620f1.png">

> the property order is different (pipeline defaults (lhs) vs node property (rhs))

Fixed! As seen in screenshot immediately above

ptitzler commented:
We should tweak the CSS. The current label padding for select lists (see highlighted areas) on top is much smaller than on the bottom. As a result the label is rendered closer to the previous element than the one that the label applies to. ![image](https://user-images.githubusercontent.com/13068832/191467615-7ef94d62-57c6-40eb-bfd1-c6915ebd3950.png)

ptitzler commented:
Manual test checklist. If checked, end-to-end test was successful:
 - proper rendering
 - proper labels
 - persistence in `.pipeline` file
 - persistence in exported DAG
 - pod properties verified

pipeline properties:
  - [ ] data volumes
  - [ ] annotations
  - [ ] environment variables
  - [ ] tolerations
  - [ ] secrets
 
node properties (KFP)
  - [ ] data volumes
  - [ ] annotations
  - [ ] environment variables
  - [ ] tolerations
  - [ ] secrets
  
node properties (Airflow)
  - [ ] data volumes
  - [ ] annotations
  - [ ] environment variables
  - [ ] tolerations
  - [ ] secrets

ptitzler commented:
Tolerations widget (node properties): missing "" in enum. the widget should render the same no matter where it is displayed (node properties vs pipeline defaults)
![image](https://user-images.githubusercontent.com/13068832/191471930-d3f1cb1b-7210-4301-9dff-19e4830f9a94.png)

ptitzler commented:
>> Since Use runtime environment default is a UI label that could change over time, it'd be safer to persist null instead to indicate that the user did not choose an explicit property value.

> I've changed this such that only True and False are listed. The empty string is persisted, which is handled during property conversion in the DisallowCachedOutput class (though full disclosure, I haven't been able to thoroughly test this due to the pipeline editor white-outs).

Observed behavior is:
 - label not shown in UI (but should be shown for clarity)
 - empty string is not persisted (which is what should make the most sense, or null)
 - property value is stored as string (`"True"`, `"False"`) but should be JSON boolean (`true`, `false`)

ptitzler commented:
Volume mount widget:
 - if no value is provided for the mount path the validation message metadata lists `/` instead of ""
   ```
   "data": {
     "nodeID": "cf9a1a8e-6d75-4c28-a9fb-41ac0ad9d6e5",
     "nodeName": "Untitled.ipynb",
     "propertyName": "mounted_volumes",
     "value": {
       "path": "/",
       "pvc_name": ""
     }
   }
   ```

ptitzler commented:
Environment variables widget :

Environment variables that are detected in the notebook/script

![image](https://user-images.githubusercontent.com/13068832/191487535-feac0171-ba74-43ab-854b-79d65ccac231.png)

are not properly populated

![image](https://user-images.githubusercontent.com/13068832/191487435-56d64c1e-3ed9-4fc4-97c7-5f7608d36e7b.png)

ptitzler commented:
Added bug fix for https://github.com/elyra-ai/elyra/pull/2927#issuecomment-1253461505 in https://github.com/elyra-ai/elyra/pull/2927/commits/d96cb11255c4f2f33ced53c39b73d0b0dcd00d0b and addressed two other issues not listed above

ptitzler commented:
Added https://github.com/elyra-ai/elyra/pull/2927/commits/1d2875ef0f8f9691b8391e01201de977487be7b4, which updates section "Node properties reference" in the pipeline documenation.

ptitzler commented:
https://github.com/elyra-ai/elyra/pull/2927 renames all occurrences and variations of "disallow cached output" to "disable node caching" in the source code (variables, methods, etc)

kevin-bates commented on elyra/pipeline/local/processor_local.py:
```diff
@@ -114,6 +115,18 @@ def process(self, pipeline):
+            pipeline_constants.KUBERNETES_POD_ANNOTATIONS,
+            pipeline_constants.KUBERNETES_TOLERATIONS,
+            pipeline_constants.DISABLE_NODE_CACHING,
+        ]
```
I think only env variables are support by the local processor.  I'm guessing this is another example of not promoting local _mode_ to be its own _processor type_, and what is really meant by this list are the properties that can be associated with Notebook, Python, and R nodes.  The `LocalPipelineProcessor` (today) only knows how to deal with envs.

kiersten-stokes replied:
Yes, that's exactly it. I had this in the PipelineProcessor before, perhaps I should move it back there for clarity?

kevin-bates replied:
I see, thanks.  I think it makes sense to leave here, but only contain `env_vars`, since others don't apply to `LocalPipelineProcessor`.

I suppose you could move to `PipelineProcessor` but only define env_vars there (since, I'm guessing, every runtime will support env variables).  Then you wouldn't need the attribute check.  If we found that some new processor didn't support envs, then we could have the base class return the empty array and move the envs_vars to `LocalPipelineProcessor` at that time - which still removes the need for the attribute check.

kiersten-stokes replied:
Well the issue is that the call to grab the "local" properties has to display all the properties available for generic nodes in any runtime type since they can be run anywhere. That's actually why, to determine the properties to render for each runtime, I'm using this "supported_properties" method instead of the class attribute check for an "add_{property}" method. Since "local" needs to pull in all these properties for which an "add" function will never be relevant, this list seemed a slightly better alternative. And, as we chatted about, we can't use the parent class to do the "add_{property}" attribute check since then there will never be a difference in properties available between runtimes.

kiersten-stokes replied:
I think I could probably do a thing where, when "local" properties are being queried, take the set of properties supported by all subclasses of 'PipelineProcessor' (or 'RuntimePipelineProcessor'). For reasons too long to explain here, this wouldn't require an explicit reference to "local" in the code

kevin-bates replied:
Thank you for the explanations.  This isn't a big enough deal to spend time on at the moment.  It's helpful to know that this is a side-effect of _not_ making "local mode" its own runtime and something we can address should we decide to promote it to its own runtime down the road.  Thanks for spending time on this.  I know things like this can be extremely frustrating.

ajbozarth commented:
I just pushed the update that add migration. If you created a pipeline on this branch previously the migration will error, but you can just update the version in the json to 8 and it should be valid

kevin-bates commented on elyra/pipeline/component_parameter.py:
```diff
@@ -0,0 +1,804 @@
+                f"'{self.operator}' is not a valid operator: the value must be one of 'Exists' or 'Equal'."
+            )
+
+        if self.operator == "Equal" and (self.key is None or len(self.key) == 0):
```
This can be cleaned up a bit using a Falsey...
```suggestion
        if self.operator == "Equal" and not self.key:
```

kevin-bates commented on elyra/pipeline/component_parameter.py:
```diff
@@ -0,0 +1,804 @@
+                f"of 'NoExecute', 'NoSchedule', or 'PreferNoSchedule'."
+            )
+
+        if self.operator == "Exists" and self.value is not None and len(self.value) > 0:
```
Ditto here...
```suggestion
        if self.operator == "Exists" and self.value:
```

ajbozarth commented:
just pushed updates to address @ptitzler issues discussed in scrum today

ptitzler commented:
> just pushed updates to address @ptitzler issues discussed in scrum today

Thank you! I confirmed 

-  the width is now consistent
  
  ![image](https://user-images.githubusercontent.com/13068832/192203077-ab2361ef-d883-4d32-a23f-b17ea62fd1e6.png)


- label padding for select lists is now more balanced

![image](https://user-images.githubusercontent.com/13068832/192203179-85918524-124b-4a00-9deb-c1337df98f1a.png)

ptitzler commented:
Currently properties for custom nodes are visually grouped into categories:
 - Inputs
 - Outputs
 - Additional properties
 - Component source
 
We should do the same for generic nodes (either in this PR or a follow-up). Doing so would provide for a more uniform user experience and prepare for the migration of the current one-off implementation to one that utilize a genuine custom component to run noteboooks and scripts.

ptitzler reviewed (approved):
LGTM! Very nice work. Gone are the days of esoteric input!

ajbozarth commented:
just pushed to update to point at the RC release of pipeline editor now that https://github.com/elyra-ai/pipeline-editor/pull/201 is merged and released
